### DISCUSSION TITLE
Performance Improvements Updates

### DISCUSSION THREAD

**Discussion by heywhy** [heywhy/ex_elasticlunr#19] [General]
This #18 is the beginning of improvements for Elasticlunr. One of the ways to benchmark these changes is by using existing data from the Elasticsearch rally project and that's why the benchmarking snippet has similar properties.

When you compare the current version (v0.6) and the new changes, you will notice very huge improvements in IPS, memory usage, and the average time taken to execute these operations. Moreover, the new changes can take on more sample sizes as opposed to the current version. 😄 

## Benchmark Track

```elixir
defmodule Elasticlunr.Benchmark.GeonamesTrack do
  use Elasticlunr.Benchmark

  alias Elasticlunr.Pipeline

  index "geonames" do
    field("elevation")
    field("name")
    field("geonameid")
    field("feature_class")
    # field("location")
    field("cc2")
    field("timezone")
    field("dem")
    field("country_code")
    field("admin1_code")
    field("admin2_code")
    field("admin4_code")
    field("feature_code")
    field("alternatenames")
    field("asciiname")
    field("population")

    pipeline(Pipeline.new(Pipeline.default_runners()))
  end

  corpora "geonames" do
    url("https://rally-tracks.elastic.co/geonames")

    document(
      document_count: 1_000,
      source_file: "documents-2-1k.json.bz2"
    )
  end

  operation("index-append", :bulk, bulk_size: 5_000, on_conflict: :ignore)

  operation("default", :search, %{
    "query" => %{
      "match_all" => %{}
    }
  })

  operation("term", :search, %{
    "query" => %{
      "terms" => %{
        "country_code" => "AT"
      }
    }
  })
end
```

### v0.6

![v0 6-elasticlunr-1](https://user-images.githubusercontent.com/13123868/156262685-db4b3b16-2978-46cf-9d5c-0f69193b2a38.jpg)

![v0 6-elasticlunr-2](https://user-images.githubusercontent.com/13123868/156262699-6e24a9db-3d5b-43ea-b77d-0bbad8bf2ad7.jpg)

### v0.7

![v0 7-elasticlunr-1](https://user-images.githubusercontent.com/13123868/156262735-1f77846a-6fa0-4201-90fb-d82090141dc6.jpg)

![v0 7-elasticlunr-2](https://user-images.githubusercontent.com/13123868/156262745-ac0b1839-87b9-4177-9153-b311a8dd9fe3.jpg)

**Comment by heywhy** [comment]
I know the code snippet looks alien, it's a benchmarking tool I'm working on, built on top of Benchee. I believe the benchmarking tool will improve over time so that we can capture important metrics as Elasticlunr improves.

**Reply by barkerja** [reply]
@heywhy this looks great. Great work! Really looking forward to learning more about the benchmarking tool.

I'm also interested in learning more about what your vision is for `ex_elasticlunr`. Particularly, do you have any plans or desires to introduce range filters?

**Reply by heywhy** [reply]
My vision for the library is to be an alternative to popular search engines so that companies or teams that don't have the resources to deploy and maintain the popular engines can have something to work with while taking advantage of the BEAM capabilities. I also look to have the library used in distributed systems.

There are plans to add more query types and also support field types so that terms can be indexed and searched accordingly. I'm looking to add more field types (`time`, `date`, `datetime`, ...) so that you can make sense of your data using their chronological attributes.

**Reply by barkerja** [reply]
Thanks for the response! I'm very interested in this project. I have a lot of experience in working with Elasticsearch, but I don't have much experience with the actual inner workings of a search engine and indexer; so I will be following along closely and hopefully learning a thing or two!
